docs/docs/higher-order-components.md

```diff
@@ -0,0 +1,421 @@
+---
+id: higher-order-components
+title: Higher-Order Components
+permalink: docs/higher-order-components.html
+---
+
+A higher-order component (HOC) is an advanced technique in React for reusing component logic. HOCs are not part of the React API, per se. They are a pattern that emerges from React's compositional nature.
+
+Concretely, **a higher-order component is a function that takes a component and returns a new component.**
+
+```js
+const EnhancedComponent = higherOrderComponent(WrappedComponent);
+```
+
+Whereas a component transforms props into UI, a higher-order component transforms themselves.
+
+HOCs are common in third-party React libraries, such as Redux's [`connect`](https://github.com/reactjs/react-redux/blob/master/docs/api.md#connectmapstatetoprops-mapdispatchtoprops-mergeprops-options) and Relay's [`createContainer`](https://facebook.github.io/relay/docs/api-reference-relay.html#createcontainer-static-method).
+
+In this document, we'll discuss why higher-order components are useful, and how to write your own.
+
+## Use HOCs For Cross-Cutting Concerns
+
+> **Note**
+>
+> We previously recommended mixins as a way to handle cross-cutting concerns. We've since realized that mixins create more trouble than they are worth. [Read more](/react/blog/2016/07/13/mixins-considered-harmful.html) about why we've moved away from mixins and how you can transition your existing components.
+
+Components are the primary unit of code reuse in React. However, you'll find that some patterns aren't a straightforward fit for traditional components.
+
+For example, say you have a `CommentList` component that subscribes to an external data source to render a list of comments:
+
+```js
+class CommentList extends React.Component {
+  constructor(props) {
+    super();
```

Comment:
I'll just remove the `props` argument entirely. Like `context`, it's not actually required.